Eager agg

iyerr3 opened the pull request:
Goal of this work is to add a transform that eagerly aggregates data if
an inner join is present below the aggregate. This allows potentially
reducing the number of tuples to join, hence making this plan cheaper
than alternatives.

Current restrictions include:
1. Only aggregates using columns from one side of the inner join are supported
2. Supported only for internal aggregates (`min`, `max`, `sum`, `count` and `avg`).
3. DQAs are not supported

Example query:
```
SELECT min(s1)
FROM foo, bar
WHERE foo.j1 = bar.j2
AND   foo.s1 % 2 = 0
GROUP BY foo.g1, bar.g2
```

Plan without EagerAgg
```
                                                  QUERY PLAN
--------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..43788.53 rows=562500 width=4)
   ->  Result  (cost=0.00..43780.14 rows=187500 width=4)
         ->  HashAggregate  (cost=0.00..43780.14 rows=187500 width=4)
               Group Key: foo.g1, bar.g2
               ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..43734.55 rows=187500 width=12)
                     Hash Key: foo.g1, bar.g2
                     ->  Result  (cost=0.00..43727.51 rows=187500 width=12)
                           ->  HashAggregate  (cost=0.00..43727.51 rows=187500 width=12)
                                 Group Key: foo.g1, bar.g2
                                 ->  Hash Join  (cost=0.00..11032.52 rows=132874743 width=12)
                                       Hash Cond: (foo.j1 = bar.j2)
                                       ->  Table Scan on foo  (cost=0.00..2595.31 rows=13333572 width=12)
                                             Filter: ((s1 % 2) = 0)
                                       ->  Hash  (cost=431.77..431.77 rows=33341 width=8)
                                             ->  Table Scan on bar  (cost=0.00..431.77 rows=33341 width=8)
```

Plan with EagerAgg (Note the `HashAggregate` below the `Hash Join`)
<pre>
                                                   QUERY PLAN
----------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..12090.26 rows=562500 width=4)
   ->  Result  (cost=0.00..12081.88 rows=187500 width=4)
         ->  HashAggregate  (cost=0.00..12081.88 rows=187500 width=4)
               Group Key: foo.g1, bar.g2
               ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..12036.29 rows=187500 width=12)
                     Hash Key: foo.g1, bar.g2
                     ->  Result  (cost=0.00..12029.25 rows=187500 width=12)
                           ->  HashAggregate  (cost=0.00..12029.25 rows=187500 width=12)
                                 Group Key: foo.g1, bar.g2
                                 ->  Hash Join  (cost=0.00..7414.04 rows=18754313 width=12)
                                       Hash Cond: (foo.j1 = bar.j2)
                                       ->  <b>HashAggregate (cost=0.00..5843.87 rows=1881938 width=12)
                                             Group Key: foo.j1, foo.g1 </b> 
                                             ->  Table Scan on foo  (cost=0.00..2595.31 rows=13333572 width=12)
                                                   Filter: ((s1 % 2) = 0)
                                       ->  Hash  (cost=431.77..431.77 rows=33341 width=8)
                                             ->  Table Scan on bar  (cost=0.00..431.77 rows=33341 width=8)
</pre>

vraghavan78 commented on libgpopt/include/gpopt/xforms/CXformEagerAgg.h:
```diff
@@ -0,0 +1,107 @@
+			BOOL FCompatible(CXform::EXformId exfid)
+			{
+				return (CXform::ExfEagerAgg != exfid) &&
+							(CXform::ExfSplitGbAgg != exfid);
```
we should probably add `ExfSplitDQA`

vraghavan78 commented on libgpopt/include/gpopt/xforms/CXformEagerAgg.h:
```diff
@@ -0,0 +1,107 @@
+        	// is this agg supported for push down?
+        	BOOL IsAggSupported (CExpression *scalar_agg_func_expr) const;
+
+			void PopulateLowerUpperProjectList
```
can u add a comment

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+    /*
+         1. check if aggregate columns is part of single child col ref
+    */
+    // Get col ref set of projection list (columns used in aggregation)
```
if this comment premature? the code below does not relate to it.

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+{
+	GPOS_ASSERT(NULL != pxf_ctxt);
+	GPOS_ASSERT(FPromising(pxf_ctxt->Pmp(), this, pexpr));
+	GPOS_ASSERT(FCheckPattern(pexpr));
```
do we ever check for aggregate with outer reference. Do we handle that?

iyerr3 replied:
Not clear on what you mean here. Could you point to an example elsewhere?

vraghavan78 replied:
```
select * from foo where foo.a = (select min(foo.b) from bar, jazz where bar.a = jazz.a group by bar.b limit 1);
```

iyerr3 replied:
This particular query does not work (since aggregates are not allowed inside `where`) but we do handle such a situation and EagerAgg will not apply if `foo.b` is not coming from `bar` or `jazz`.

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+    /*   2. get grouping columns and join predicate columns into a new set  */
+    CLogicalGbAgg *gb_agg_op = CLogicalGbAgg::PopConvert(pexpr->Pop());
+    CColRefSet *grouping_crs = gb_agg_op->PcrsLocalUsed();
+    CColRefSet *push_down_gb_crs = GPOS_NEW(mp) CColRefSet
```
Can you add a detailed comment here, about what and why you are doing this?

vraghavan78 replied:
consider the example:

```sql
 select min(a1) from t1, t2 where t1.j1 = t2.j2 group by t1.g1
```

We can to convert the above query into 

```
select min(pushed-min1)
from 
(select min(s1) as pushed-min1 from t1 group by t1.g1, t1.j1) x,
t2
where t1.j1 = t2.j2
group by t1.g1
```

In the above example, we need to collect `t1.j1` and `t1.g1` as the grouping column for the new group by `min(s1) as pushed-min1` on top the join child `t1`

@iyerr3 what do you think about the example. Wouldn't help the reader?

iyerr3 replied:
Added the example to the code.

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+    CExpression *proj_list_expr = (*pexpr)[1];
+    CExpression *outer_child_expr = (*join_expr)[0];
+    CExpression *inner_child_expr = (*join_expr)[1];
+    CExpression *scalar_expr = (*join_expr)[2];
```
rename join_condition_expr

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+    // Get col ref set of projection list (columns used in aggregation)
+    CExpression *join_expr = (*pexpr)[0];
+    CExpression *proj_list_expr = (*pexpr)[1];
+    CExpression *outer_child_expr = (*join_expr)[0];
```
join_outer_child_expr

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+    CExpression *join_expr = (*pexpr)[0];
+    CExpression *proj_list_expr = (*pexpr)[1];
+    CExpression *outer_child_expr = (*join_expr)[0];
+    CExpression *inner_child_expr = (*join_expr)[1];
```
join_inner_child_expr

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+    */
+    // Get col ref set of projection list (columns used in aggregation)
+    CExpression *join_expr = (*pexpr)[0];
+    CExpression *proj_list_expr = (*pexpr)[1];
```
agg_proj_list_expr

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+//     - Single expression input in the agg
+//     - Input expression only part of outer child
+BOOL
+CXformEagerAgg::FApplicable
```
I think the non-aggregation is "CanApplyTransform" :)

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+BOOL
+CXformEagerAgg::FApplicable
+(
+ CExpression *pexpr,
```
pexprAgg

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+const
+{
+	CExpression *join_expr = (*pexpr)[0];
+	CExpression *proj_list_expr = (*pexpr)[1];
```
agg_proj_list_expr

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+                                   outer_child_expr->PdpDerive())->PcrsOutput();
+    CColRefSet *proj_list_crs = CDrvdPropScalar::GetDrvdScalarProps(
+                                proj_list_expr->PdpDerive())->PcrsUsed();
+    if (!outer_child_crs->ContainsAll(proj_list_crs))
```
can you add an example of what this is doing?

iyerr3 replied:
added comment to describe the comparison.

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+	CExpression *outer_child_expr = (*join_expr)[0];
+
+    // currently only supporting aggregate column references from outer child //
+    CColRefSet *outer_child_crs = CDrvdPropRelational::GetRelationalProperties(
```
join_outer_child_crs

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+    // currently only supporting aggregate column references from outer child //
+    CColRefSet *outer_child_crs = CDrvdPropRelational::GetRelationalProperties(
+                                   outer_child_expr->PdpDerive())->PcrsOutput();
+    CColRefSet *proj_list_crs = CDrvdPropScalar::GetDrvdScalarProps(
```
agg_proj_list_crs

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+        return false;
+    }
+    BOOL isSupported = false;
+	ULONG arity = proj_list_expr->Arity();
```
const

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+    }
+    for (ULONG ul = 0; ul < arity; ul++)
+    {
+        // currently only supporting single-input aggregates //
```
no ending //

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+        // currently only supporting single-input aggregates //
+		// scalar aggregate function is the lone child of the project element
+
+        // currently only supporting MIN/MAX aggregate function //
```
same here. Instead use /* */

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+    for (ULONG ul = 0; ul < arity; ul++)
+    {
+        // currently only supporting single-input aggregates //
+		// scalar aggregate function is the lone child of the project element
```
is it? I thought u support multiple aggregate

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+        // currently only supporting MIN/MAX aggregate function //
+		CExpression *scalar_agg_proj_expr = (*proj_list_expr)[ul];
+        BOOL is_agg_supported = IsAggSupported((*scalar_agg_proj_expr)[0]);
+        if (is_agg_supported)
```
inline the function call in the IF clause

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+        } else
+        {
+            // An unsupported agg can only be applied on the grouping columns
+            // of the aggregate being pushed down
```
can you show an example of what you support and what you do not and why

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+	pxfres->Add(upper_agg_expr);
+}
+
+BOOL CXformEagerAgg::IsAggSupported
```
CanPushAggBelowJoin

vraghavan78 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,413 @@
+    if (scalar_agg_func_expr->Arity() != 1)
+    {
+        // currently count() is not supported since it's not possible to
+        // distinguish it from a dummy aggregate with no child expression
```
not only count but also User Defined Aggregates correct. If so fix comment.

dhanashreek89 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,424 @@
+const
+{
+	CExpression *join_expr = (*gb_agg_pexpr)[0];
+	CExpression *agg_proj_list_expr = (*gb_agg_pexpr)[1];
```
- Can we perform this transformation if the join is not an equijoin ? If no, then where do we check for supported join conditions?

- Do cross joins need to be handled specially? Specifically while generating grouping keys for pushed down agg.  Also wondering if it is worthwhile to perform eager aggregation for cross joins ?

hardikar replied:
CLogicalInnerJoin is explicitly used in the xform patter, that I think takes care of that. @iyerr3 can you confirm? I think having an assert for that is also harmless and clarifies that intention. Let's add one.

hardikar replied:
CLogicalInnerJoin is explicitly used in the xform patter, that I think takes care of that. @iyerr3 can you confirm? I think having an assert for that is also harmless and clarifies that intention. Let's add one.

iyerr3 replied:
@hardikar I believe the `GPOS_ASSERT(FCheckPattern(pexpr));` in `Transform()` checks this (+ other things). Would you suggest another assert for documentation purposes?

bhuvnesh2703 replied:
i think we don't need it. FCheckPattern is itself in an assert, and while it goes to check the pattern for the children it compares the operator id..(including inner join as thats in pattern)

hardikar commented on libgpopt/src/operators/CPhysicalHashAgg.cpp:
```diff
@@ -52,9 +53,12 @@ CPhysicalHashAgg::CPhysicalHashAgg
 	)
-{}
+{
+	m_should_enforce_distribution = should_enforce_distribution;
```
Why is this double assignment needed?

iyerr3 replied:
Not needed. removing ...

hardikar commented on libgpopt/include/gpopt/operators/CPhysicalHashAgg.h:
```diff
@@ -47,7 +50,8 @@ namespace gpopt
 				BOOL isAggFromSplitDQA,
-				CLogicalGbAgg::EAggStage aggStage
+				CLogicalGbAgg::EAggStage aggStage,
+				BOOL should_enforce_distribution = true
```
Is `should_enforce_distribution` ever false? If so, can you add a comment where it's initialized as true, and i not, let's remove this altogether

iyerr3 replied:
The flag is set to `false` when `CXformEagerAgg` is used - hence checked in `CXformGbAgg2*Agg` transforms. I've added a comment to that effect.

hardikar commented on data/dxl/minidump/EagerAggExpression.mdp:
```diff
@@ -0,0 +1,3527 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I'm not getting the same plan as this MDP:
```
shardikar=#        explain  select min(s1%2)
shardikar-#        from foo, bar
shardikar-#        where j1 = j2 and s1%2 = 0
shardikar-#        group by g1;
                                                QUERY PLAN
-----------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..25602.94 rows=1001 width=4)
   ->  Result  (cost=0.00..25602.92 rows=334 width=4)
         ->  HashAggregate  (cost=0.00..25602.92 rows=334 width=4)
               Group Key: foo.g1
               ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..25602.88 rows=334 width=8)
                     Hash Key: foo.g1
                     ->  Result  (cost=0.00..25602.88 rows=334 width=8)
                           ->  HashAggregate  (cost=0.00..25602.88 rows=334 width=8)
                                 Group Key: foo.g1
                                 ->  Hash Join  (cost=0.00..9182.27 rows=133349047 width=8)
                                       Hash Cond: (foo.j1 = bar.j2)
                                       ->  Table Scan on foo  (cost=0.00..2595.31 rows=13333572 width=12)
                                             Filter: ((s1 % 2) = 0)
                                       ->  Hash  (cost=431.77..431.77 rows=33334 width=4)
                                             ->  Table Scan on bar  (cost=0.00..431.77 rows=33334 width=4)
 Planning time: 3259.628 ms
 Optimizer: PQO version 3.13.0
(17 rows)
```
Am I missing some GUC or some setting?

iyerr3 replied:
Try with `set optimizer_experimental to on`. It can still end up picking this plan if the data is different.

iyerr3 commented:
@hardikar Thanks for the initial PR. I'll make the updates and comment individual change requests. 

- I was unsure of the variable naming and didn't find a guideline document. Do you suggest not adding `CRS/CRA` in any of the names? 
- The GPDB PR has not been created yet - that contains the `optimizer_experimental` GUC ([diff](https://github.com/greenplum-db/gpdb/compare/master...iyerr3:eager-agg?expand=1), [branch](https://github.com/iyerr3/gpdb/tree/eager-agg)).

iyerr3 commented on libgpopt/include/gpopt/xforms/CXformEagerAgg.h:
```diff
@@ -0,0 +1,128 @@
+	//      Eagerly push aggregates below join when there is no primary/foreign keys
+	//
+	//---------------------------------------------------------------------------
+	class CXformEagerAgg : public CXformExploration
```
I imagined the push-down of `count` on the other side as well as pushing down an agg to both sides of a join to be extensions of the same xform. So from that perspective, I would say let's keep this name and update if we end up changing the design.

iyerr3 commented on libgpopt/include/gpopt/operators/CPhysicalHashAgg.h:
```diff
@@ -31,6 +31,9 @@ namespace gpopt
 		private:
 
+			// should we enforce distribution or not for this operator
+			BOOL m_should_enforce_distribution;
```
I'm not sure of the original intention. 
@asubramanya Could you please comment on this question?

asubramanya replied:
I agree. I think this can be removed from the subclass and retained only in the base class.

iyerr3 replied:
This is set to false in multiple `...Gb2*Agg` transforms i.e. every where below function is used:
```
CXformUtils::FLocalAggCreatedByEagerAggXform(pexpr)
```

iyerr3 commented on libgpopt/include/gpopt/operators/CLogicalGbAgg.h:
```diff
@@ -332,6 +337,8 @@ namespace gpopt
 			EAggStage m_aggStage;
 
+			// is the group by columns for local different from the global agg
+			BOOL m_has_different_local_gb_crs;
```
Removed

iyerr3 commented on libgpopt/include/gpopt/operators/CLogicalGbAgg.h:
```diff
@@ -295,11 +296,15 @@ namespace gpopt
 			}
 
-		
+			BOOL IsLocalAndDifferentGroupingCRS() const
```
Removed

hardikar commented:
As a side note, did we run this (with the GUC enabled) on the entire MDP test suite? Were there any alarming changes there?

iyerr3 commented on libgpopt/src/operators/CLogicalGbAgg.cpp:
```diff
@@ -661,6 +661,10 @@ CLogicalGbAgg::PxfsCandidates
 	(void) xform_set->ExchangeSet(CXform::ExfGbAgg2StreamAgg);
 	(void) xform_set->ExchangeSet(CXform::ExfGbAgg2ScalarAgg);
 
+	if (GPOS_FTRACE(EopttraceEnableExperiment)){
```
Makes sense - updated code. Only point in favor of this would be that if we end up with multiple experimental transforms then we have the switch for all of them in one place. We can cross that bridge (if and) when we get there.

iyerr3 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,469 @@
+	CExpression *lower_agg_expr = GPOS_NEW(mp) CExpression
+		(
+		mp,
+		GPOS_NEW(mp) CLogicalGbAgg
```
To fully understand: you're suggesting we add another bool argument within `CLogicalGbAgg` constructor and set it to True here?

iyerr3 replied:
Same as the comment above: 

> This is similar to the approach used for MultiStageAgg - the idea is that the origin of a logical agg expression is stored, precluding the need for another flag/enum. We can use that to find if an agg is obtained from CXformEagerAgg. Further we iterate through the group expression all the way to the root to ensure that we don't miss out on an EagerAgg output that got transformed again by another CXform.

iyerr3 commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,469 @@
+		CScalarProjectElement::PopConvert(orig_proj_elem_expr->Pop());
+
+		CExpression *orig_agg_expr = (*orig_proj_elem_expr)[0];
+		if (CanPushAggBelowJoin(orig_agg_expr))
```
You're right - this is a no-op check carried over from earlier work. Removed.

iyerr3 commented on libgpopt/src/xforms/CXformUtils.cpp:
```diff
@@ -1898,6 +1898,36 @@ CXformUtils::FMultiStageAgg
+	while (NULL != pgexprOrigin && !is_eager_agg)
+	{
+		// parse all expressions in group to check if any was created by CXformEagerAgg
+		is_eager_agg = CXform::ExfEagerAgg == pgexprOrigin->ExfidOrigin();
```
This is similar to the approach used for `MultiStageAgg` - the idea is that the origin of a logical agg expression is stored. We can use that to find if an agg is obtained from `CXformEagerAgg`. Further we iterate through the group expression all the way to the root to ensure that we don't miss out on an EagerAgg output that got transformed again by another CXform.

iyerr3 commented on libgpopt/src/operators/CPhysicalStreamAgg.cpp:
```diff
@@ -42,7 +42,8 @@ CPhysicalStreamAgg::CPhysicalStreamAgg
 	BOOL isAggFromSplitDQA,
-	CLogicalGbAgg::EAggStage aggStage
+	CLogicalGbAgg::EAggStage aggStage,
+  BOOL should_enforce_distribution
```
> As a side note, did we run this (with the GUC enabled) on the entire MDP test suite? Were there any alarming changes there?

We did run the whole test suite before the GUC was added in. There were some MDP changes in the earlier version of this PR - most changes were just plan size increases, while some were changes in plans, but only improvements in cost. We had spent considerable time in evaluating each and decided that all were positive changes. Those have been reverted now since the transform is disabled by default.

iyerr3 commented:
> As a side note, did we run this (with the GUC enabled) on the entire MDP test suite? Were there any alarming changes there?

We did run the whole test suite before the GUC was added in. There were some MDP changes in the earlier version of this PR - most changes were just plan size increases, while some were changes in plans, but only improvements in cost. We had spent considerable time in evaluating each and decided that all were positive changes. Those have been reverted now since the transform is disabled by default.

iyerr3 commented:
Thanks for the comments, @hardikar. Please see response below: 

> I'm still not convinced as to why we need m_should_enforce_distribution. Adding more comments would definitely help. It just seems like a code smell to me, something that might work - but could be implemented differently. I'd love to proven wrong. (Note this isn't necessarily deal breaker, let's just document the reason why it was needed in a lot of detail)

I've added a comment in `CPhysicalAgg.h` that creates the variable re: the need for the variable. (also quoted below):                     

```
// should distribution enforcement be enabled on this agg?
// By default, global and local aggregate are created with same
// grouping columns. In such cases, if local derives the same
// distribution as global then we need no motion in-between, which
// implies that a single aggregate is enough. Hence, such plans are
// prohibited. In CXformEagerAgg, however, the local agg is created
// with different grouping columns but can have the same
// distribution as the global. We don't need to prohibit such plans,
// since the global agg is applied with different grouping columns 
// compared to to the local and is still necessary.
```

> 
> Here some other test cases I could think of that might play with the code changes.
> For the tables:
> 
> create table foo (j1 int, g1 int, s1 int);
> create table bar (j2 int, g2 int, s2 int);
> g1/g2 are in the join condition: select g1, g2, max(s1) from foo inner join bar on (g1 = j2) where s1%2 = 0 group by g1, g2;
> Subqueries-1: select g1, g2, sum(s1) from foo inner join bar on (j1 = j2) where g1 IN (select j1 from bar) group by g1, g2;
> Subqueries-2: select g1, g2, (select sum(s1) from foo group by g1, g2) from foo inner join bar on (j1 = j2);

The first two do apply to eager agg - I'll add them to our current suite. The third one does not apply since the incoming expression to EagerAgg does not match the pattern we expect (i.e. aggregate over innerjoin).

iyerr3 commented:
Forced-push this to combine all commits in to one merge-able commit. 
Recent push also includes following work with @bhuvnesh2703: 
- Updated GUC name to `optimizer_enable_eageragg`
- Renamed some variable names to avoid Hungarian conventions
- Added two new test cases
- Added more comments to describe the work with `m_should_enforce_distribution`

d commented on libgpopt/src/xforms/CXformEagerAgg.cpp:
```diff
@@ -0,0 +1,454 @@
+		agg_mdid->Equals(agg_child_type->GetMdidForAggType(IMDType::EaggSum)) ||
+		agg_mdid->Equals(agg_child_type->GetMdidForAggType(IMDType::EaggCount)) ||
+		agg_mdid->Equals(agg_child_type->GetMdidForAggType(IMDType::EaggAvg))))
+	{
```
this is very strange, not because we _only_ allow `min`, `max`, `sum`, `count`, and `avg`, but because, who says those 5 will work? What happens when the `max` aggregate doesn't support splitting (i.e. `pg_aggregate.combinefn` is `InvalidOid`)?

hardikar replied:
Yeah - I don't know why max wouldn't have a combinefn, but it wouldn't hurt checking for that. Does the SplitAgg xform do any such checking, actually?
